File: src-web/src/store/websocket.ts
import { create } from "zustand"
import { capPush } from "@/lib/boundedArray"
import { useEnvironmentStore } from "@/store/environment"
import { useRequestStore } from "@/store/requests"
import { useUiStore } from "@/store/workspace"
import { resolveInheritedAuth } from "@/views/ApiWorkspace/sendResolution/inheritance"
import type {
  AuthConfig,
  StoredWsSession,
  StoredWsSessionSummary,
  TimelineEvent,
  WsMessage,
  WsMessageKind,
} from "../../../packages/types/bindings"
import { commands } from "../../../packages/types/bindings"

export type WsConnStatus =
  | "closed"
  | "connecting"
  | "open"
  | "closing"
  | "error"

interface WebsocketStore {
  status: Record<string, WsConnStatus>
  transcripts: Record<string, WsMessage[]>
  timelines: Record<string, TimelineEvent[]>

  setStatus: (connectionId: string, status: WsConnStatus) => void
  appendMessage: (connectionId: string, message: WsMessage) => void
  appendTimeline: (connectionId: string, event: TimelineEvent) => void

  /** Load the persisted transcript + current connection state for a connection. */
  hydrate: (workspaceId: string, connectionId: string) => Promise<void>
  connect: (workspaceId: string, connectionId: string) => Promise<void>
  disconnect: (workspaceId: string, connectionId: string) => Promise<void>
  sendMessage: (
    workspaceId: string,
    connectionId: string,
    kind: WsMessageKind,
    data: string,
  ) => Promise<void>
  clearTranscript: (workspaceId: string, connectionId: string) => Promise<void>
  /** List past sessions (newest first) for the connection's history picker. */
  listSessions: (
    workspaceId: string,
    connectionId: string,
  ) => Promise<StoredWsSessionSummary[]>
  /** Fetch a past session's full transcript. */
  getSession: (
    workspaceId: string,
    connectionId: string,
    sessionId: string,
  ) => Promise<StoredWsSession | null>
}

function activeEnvId(): string | null {
  return useEnvironmentStore.getState().activeEnvId
}

/** Returns a resolved auth to pass as `authOverride` to `ws_connect`, or null
 *  when no override is needed (the auth isn't Inherit, or the connection /
 *  workspace can't be found — backend falls back to stored auth). */
function resolveAuthForConnect(
  workspaceId: string,
  connectionId: string,
): AuthConfig | null {
  const connections = useRequestStore.getState().connections
  const connection = connections.find((c) => c.id === connectionId)
  if (!connection || connection.auth?.kind !== "inherit") return null
  const workspace = useUiStore
    .getState()
    .workspaces.find((w) => w.id === workspaceId)
  if (!workspace) return null
  const folders = useRequestStore.getState().folders
  return resolveInheritedAuth(connection, folders, workspace)
}

export const useWebsocketStore = create<WebsocketStore>((set, get) => ({
  status: {},
  transcripts: {},
  timelines: {},

  setStatus: (connectionId, status) =>
    set((s) =>
      s.status[connectionId] === status
        ? s
        : { status: { ...s.status, [connectionId]: status } },
    ),

  appendMessage: (connectionId, message) =>
    set((s) => {
      const current = s.transcripts[connectionId] ?? []
      // Dedup by id — the backend emits one event per message; guard re-delivery.
      if (current.some((m) => m.id === message.id)) return s
      return {
        transcripts: {
          ...s.transcripts,
          [connectionId]: capPush(current, message),
        },
      }
    }),

  appendTimeline: (connectionId, event) =>
    set((s) => {
      const current = s.timelines[connectionId] ?? []
      return {
        timelines: {
          ...s.timelines,
          [connectionId]: capPush(current, event),
        },
      }
    }),

  hydrate: async (workspaceId, connectionId) => {
    const [transcriptRes, connectedRes] = await Promise.all([
      commands.wsGetTranscript(workspaceId, connectionId),
      commands.wsIsConnected(connectionId),
    ])
    const connected = connectedRes.status === "ok" && connectedRes.data
    set((s) => ({
      transcripts: {
        ...s.transcripts,
        [connectionId]:
          transcriptRes.status === "ok"
            ? (transcriptRes.data.messages ?? [])
            : [],
      },
      timelines: {
        ...s.timelines,
        [connectionId]:
          transcriptRes.status === "ok"
            ? (transcriptRes.data.events ?? [])
            : [],
      },
      status: { ...s.status, [connectionId]: connected ? "open" : "closed" },
    }))
  },

  connect: async (workspaceId, connectionId) => {
    get().setStatus(connectionId, "connecting")
    // Each connect starts a new history session — reset live view.
    set((s) => ({
      transcripts: { ...s.transcripts, [connectionId]: [] },
      timelines: { ...s.timelines, [connectionId]: [] },
    }))
    const authOverride = resolveAuthForConnect(workspaceId, connectionId)
    const res = await commands.wsConnect(
      workspaceId,
      connectionId,
      activeEnvId(),
      authOverride,
    )
    // Status transitions otherwise arrive via `ws:status` events; on a hard
    // failure no event fires, so reflect the error here.
    if (res.status !== "ok") get().setStatus(connectionId, "error")
  },

  disconnect: async (workspaceId, connectionId) => {
    get().setStatus(connectionId, "closing")
    await commands.wsDisconnect(workspaceId, connectionId)
  },

  sendMessage: async (workspaceId, connectionId, kind, data) => {
    // No optimistic append — the backend emits a `ws:message` event for the
    // outbound row too, so the event handler is the single source of truth.
    await commands.wsSendMessage(
      workspaceId,
      connectionId,
      kind,
      data,
      activeEnvId(),
    )
  },

  clearTranscript: async (workspaceId, connectionId) => {
    await commands.wsClearTranscript(workspaceId, connectionId)
    set((s) => ({
      transcripts: { ...s.transcripts, [connectionId]: [] },
      timelines: { ...s.timelines, [connectionId]: [] },
    }))
  },

  listSessions: async (workspaceId, connectionId) => {
    const res = await commands.wsListSessions(workspaceId, connectionId)
    return res.status === "ok" ? res.data : []
  },

  getSession: async (workspaceId, connectionId, sessionId) => {
    const res = await commands.wsGetSession(
      workspaceId,
      connectionId,
      sessionId,
    )
    return res.status === "ok" ? res.data : null
  },
}))

File: src-tauri/src/commands/websocket.rs
//! WebSocket CRUD + live-session commands. Mirrors `commands/request.rs` for
//! CRUD; lifecycle commands drive `WsManager` and push `ws:*` events.

use std::sync::Arc;

use serde_json::json;
use tauri::{AppHandle, Emitter, State};
use voleeo_core::{
    new_id, now_iso, AuthConfig, RequestParameter, TimelineEvent, VoleeoError, WsConnection,
    WsMessage, WsMessageKind,
};
use voleeo_storage::{StoredWsSession, StoredWsSessionSummary};
use voleeo_ws::{WsEvent, WsEventSink};

use crate::commands::request::{
    preserve_unchanged_secrets, run_blocking, transform_auth_secrets, Direction, Stores,
};
use crate::state::AppState;

#[tauri::command]
#[specta::specta]
pub async fn list_ws_connections(
    state: State<'_, AppState>,
    workspace_id: String,
) -> Result<Vec<WsConnection>, VoleeoError> {
    let stores = Stores::from(&state);
    let ws = state.ws.clone();
    run_blocking(move || {
        let mut conns = ws.list(&workspace_id)?;
        for c in conns.iter_mut() {
            transform_auth_secrets(&mut c.auth, &workspace_id, &stores, Direction::Decrypt)?;
        }
        Ok(conns)
    })
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn get_ws_connection(
    state: State<'_, AppState>,
    workspace_id: String,
    id: String,
) -> Result<WsConnection, VoleeoError> {
    let stores = Stores::from(&state);
    let ws = state.ws.clone();
    run_blocking(move || {
        let mut conn = ws.get(&workspace_id, &id)?;
        transform_auth_secrets(&mut conn.auth, &workspace_id, &stores, Direction::Decrypt)?;
        Ok(conn)
    })
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn create_ws_connection(
    state: State<'_, AppState>,
    workspace_id: String,
    folder_id: Option<String>,
    name: String,
    url: String,
) -> Result<WsConnection, VoleeoError> {
    let ws = state.ws.clone();
    run_blocking(move || ws.create(workspace_id, folder_id, name, url)).await
}

#[tauri::command]
#[specta::specta]
pub async fn duplicate_ws_connection(
    state: State<'_, AppState>,
    workspace_id: String,
    id: String,
) -> Result<WsConnection, VoleeoError> {
    let stores = Stores::from(&state);
    let ws = state.ws.clone();
    run_blocking(move || {
        let mut conn = ws.duplicate(&workspace_id, &id)?;
        transform_auth_secrets(&mut conn.auth, &workspace_id, &stores, Direction::Decrypt)?;
        Ok(conn)
    })
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn rename_ws_connection(
    state: State<'_, AppState>,
    workspace_id: String,
    id: String,
    name: String,
) -> Result<(), VoleeoError> {
    let ws = state.ws.clone();
    run_blocking(move || ws.rename(&workspace_id, &id, name)).await
}

#[tauri::command]
#[specta::specta]
#[allow(clippy::too_many_arguments)]
pub async fn update_ws_connection(
    state: State<'_, AppState>,
    workspace_id: String,
    id: String,
    url: String,
    parameters: Vec<RequestParameter>,
    headers: Vec<RequestParameter>,
    auth: AuthConfig,
) -> Result<(), VoleeoError> {
    let stores = Stores::from(&state);
    let ws = state.ws.clone();
    run_blocking(move || {
        // Plaintext compare before encrypting — see `update_request`.
        let mut current = ws.get(&workspace_id, &id)?;
        let mut stored_auth = current.auth.clone();
        transform_auth_secrets(
            &mut current.auth,
            &workspace_id,
            &stores,
            Direction::Decrypt,
        )?;
        let mut next = current.clone();
        next.url = url;
        next.parameters = parameters;
        next.headers = headers;
        next.auth = auth;
        if next == current {
            return Ok(());
        }
        preserve_unchanged_secrets(&mut next.auth, &mut current.auth, &mut stored_auth);
        transform_auth_secrets(&mut next.auth, &workspace_id, &stores, Direction::Encrypt)?;
        ws.update(
            &workspace_id,
            &id,
            next.url,
            next.parameters,
            next.headers,
            next.auth,
        )
    })
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn delete_ws_connection(
    state: State<'_, AppState>,
    workspace_id: String,
    id: String,
) -> Result<(), VoleeoError> {
    let ws = state.ws.clone();
    let transcripts = state.ws_transcripts.clone();
    let conn_id = id.clone();
    let ws_id = workspace_id.clone();
    run_blocking(move || {
        let _ = transcripts.clear(&ws_id, &conn_id);
        ws.delete(&workspace_id, &id)
    })
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn ws_update_position(
    state: State<'_, AppState>,
    workspace_id: String,
    id: String,
    folder_id: Option<String>,
    order: f64,
) -> Result<(), VoleeoError> {
    let ws = state.ws.clone();
    run_blocking(move || ws.update_position(&workspace_id, &id, folder_id, order)).await
}

/// Build the sink the manager calls for every inbound frame / lifecycle event:
/// emit a small `ws:*` event to the frontend and persist to the transcript.
fn build_sink(
    app: AppHandle,
    state: &AppState,
    workspace_id: String,
    connection_id: String,
) -> WsEventSink {
    let transcripts = state.ws_transcripts.clone();
    Arc::new(move |ev: WsEvent| match ev {
        WsEvent::Status(status) => {
            let _ = app.emit(
                "ws:status",
                json!({ "connectionId": connection_id, "status": status }),
            );
        }
        WsEvent::Message(msg) => {
            let _ = app.emit(
                "ws:message",
                json!({ "connectionId": connection_id, "message": &msg }),
            );
            let (t, ws, c) = (
                transcripts.clone(),
                workspace_id.clone(),
                connection_id.clone(),
            );
            tauri::async_runtime::spawn_blocking(move || {
                let _ = t.append_message(&ws, &c, msg);
            });
        }
        WsEvent::Timeline(event) => {
            let _ = app.emit(
                "ws:timeline",
                json!({ "connectionId": connection_id, "event": &event }),
            );
            let (t, ws, c) = (
                transcripts.clone(),
                workspace_id.clone(),
                connection_id.clone(),
            );
            tauri::async_runtime::spawn_blocking(move || {
                let _ = t.append_event(&ws, &c, event);
            });
        }
    })
}

/// Resolves env/folder vars + auth backend-side; `auth_override` short-circuits
/// the stored auth (frontend uses it for already-walked `Inherit`).
#[tauri::command]
#[specta::specta]
pub async fn ws_connect(
    app: AppHandle,
    state: State<'_, AppState>,
    workspace_id: String,
    connection_id: String,
    environment_id: Option<String>,
    auth_override: Option<AuthConfig>,
) -> Result<(), VoleeoError> {
    let stores = Stores::from(&state);
    let ws = state.ws.clone();
    let environments = state.environments.clone();
    let requests = state.requests.clone();
    let app_data_dir = state.app_data_dir.clone();
    let (ws_id, conn_id, env_id) = (
        workspace_id.clone(),
        connection_id.clone(),
        environment_id.clone(),
    );

    let transcripts = state.ws_transcripts.clone();
    let (url, headers) = run_blocking(move || {
        let mut conn = ws.get(&ws_id, &conn_id)?;
        if let Some(auth) = auth_override {
            // Frontend already resolved inheritance; trust it. (Secrets in the
            // override travel plaintext over IPC, same as `request.send`.)
            conn.auth = auth;
        } else {
            transform_auth_secrets(&mut conn.auth, &ws_id, &stores, Direction::Decrypt)?;
        }
        let envs = environments.list(&ws_id).unwrap_or_default();
        let mut vars = voleeo_mcp::resolve::load_env_vars_from(
            &envs,
            &ws_id,
            env_id.as_deref(),
            &app_data_dir,
        );
        let folders = requests.list_folders(&ws_id).unwrap_or_default();
        let key = voleeo_crypto::load_key_from_file(&ws_id, &app_data_dir).ok();
        voleeo_mcp::resolve::apply_folder_vars(
            &mut vars,
            conn.folder_id.as_deref(),
            &folders,
            key.as_ref(),
        );
        let resolved = voleeo_mcp::resolve::apply_to_connection(&conn, &vars);
        // Open a fresh history session so handshake/open events + messages from
        // this connect land in their own entry.
        let _ = transcripts.start_session(&ws_id, &conn_id);
        Ok::<(String, Vec<(String, String)>), VoleeoError>(resolved)
    })
    .await?;

    let sink = build_sink(app, &state, workspace_id, connection_id.clone());
    state
        .ws_manager
        .connect(connection_id, url, headers, sink)
        .await
}

/// Send a message (optimistic): `{{ VAR }}` tokens in text payloads are resolved,
/// then the outbound row is emitted + persisted immediately and enqueued.
#[tauri::command]
#[specta::specta]
pub async fn ws_send_message(
    app: AppHandle,
    state: State<'_, AppState>,
    workspace_id: String,
    connection_id: String,
    kind: WsMessageKind,
    data: String,
    environment_id: Option<String>,
) -> Result<WsMessage, VoleeoError> {
    // Binary payloads are base64 — leave untouched; resolve text only.
    let resolved = if matches!(kind, WsMessageKind::Text) {
        let ws = state.ws.clone();
        let environments = state.environments.clone();
        let requests = state.requests.clone();
        let app_data_dir = state.app_data_dir.clone();
        let (ws_id, conn_id, env_id, raw) = (
            workspace_id.clone(),
            connection_id.clone(),
            environment_id.clone(),
            data.clone(),
        );
        run_blocking(move || {
            let conn = ws.get(&ws_id, &conn_id)?;
            let envs = environments.list(&ws_id).unwrap_or_default();
            let mut vars = voleeo_mcp::resolve::load_env_vars_from(
                &envs,
                &ws_id,
                env_id.as_deref(),
                &app_data_dir,
            );
            let folders = requests.list_folders(&ws_id).unwrap_or_default();
            let key = voleeo_crypto::load_key_from_file(&ws_id, &app_data_dir).ok();
            voleeo_mcp::resolve::apply_folder_vars(
                &mut vars,
                conn.folder_id.as_deref(),
                &folders,
                key.as_ref(),
            );
            Ok(voleeo_mcp::resolve::resolve_str(&raw, &vars))
        })
        .await?
    } else {
        data.clone()
    };

    state
        .ws_manager
        .send_message(&connection_id, kind, resolved.clone())?;

    let msg = WsMessage {
        id: new_id(),
        direction: voleeo_core::WsDirection::Outgoing,
        kind,
        data: resolved.clone(),
        size: resolved.len() as u32,
        at: now_iso(),
    };
    let _ = app.emit(
        "ws:message",
        json!({ "connectionId": connection_id, "message": &msg }),
    );
    let transcripts = state.ws_transcripts.clone();
    let (ws, c, stored) = (workspace_id, connection_id, msg.clone());
    tauri::async_runtime::spawn_blocking(move || {
        let _ = transcripts.append_message(&ws, &c, stored);
    });
    Ok(msg)
}

#[tauri::command]
#[specta::specta]
pub async fn ws_disconnect(
    app: AppHandle,
    state: State<'_, AppState>,
    workspace_id: String,
    connection_id: String,
) -> Result<(), VoleeoError> {
    state.ws_manager.disconnect(&connection_id);
    let _ = app.emit(
        "ws:status",
        json!({ "connectionId": connection_id, "status": "closed" }),
    );
    let event = TimelineEvent {
        at_ms: 0.0,
        kind: "close".into(),
        text: "Disconnected".into(),
    };
    let _ = app.emit(
        "ws:timeline",
        json!({ "connectionId": connection_id, "event": &event }),
    );
    let transcripts = state.ws_transcripts.clone();
    tauri::async_runtime::spawn_blocking(move || {
        let _ = transcripts.append_event(&workspace_id, &connection_id, event);
    });
    Ok(())
}

#[tauri::command]
#[specta::specta]
pub async fn ws_is_connected(
    state: State<'_, AppState>,
    connection_id: String,
) -> Result<bool, VoleeoError> {
    Ok(state.ws_manager.is_connected(&connection_id))
}

/// The current (newest) session — used to hydrate the live transcript on mount.
#[tauri::command]
#[specta::specta]
pub async fn ws_get_transcript(
    state: State<'_, AppState>,
    workspace_id: String,
    connection_id: String,
) -> Result<StoredWsSession, VoleeoError> {
    let transcripts = state.ws_transcripts.clone();
    run_blocking(move || Ok(transcripts.latest(&workspace_id, &connection_id))).await
}

/// Session history (newest first) for the connection's history picker.
#[tauri::command]
#[specta::specta]
pub async fn ws_list_sessions(
    state: State<'_, AppState>,
    workspace_id: String,
    connection_id: String,
) -> Result<Vec<StoredWsSessionSummary>, VoleeoError> {
    let transcripts = state.ws_transcripts.clone();
    run_blocking(move || Ok(transcripts.list_sessions(&workspace_id, &connection_id))).await
}

/// A single past session's full transcript + lifecycle events.
#[tauri::command]
#[specta::specta]
pub async fn ws_get_session(
    state: State<'_, AppState>,
    workspace_id: String,
    connection_id: String,
    session_id: String,
) -> Result<StoredWsSession, VoleeoError> {
    let transcripts = state.ws_transcripts.clone();
    run_blocking(move || {
        transcripts
            .get_session(&workspace_id, &connection_id, &session_id)
            .ok_or_else(|| VoleeoError::NotFound(format!("session {session_id}")))
    })
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn ws_clear_transcript(
    state: State<'_, AppState>,
    workspace_id: String,
    connection_id: String,
) -> Result<(), VoleeoError> {
    let transcripts = state.ws_transcripts.clone();
    run_blocking(move || transcripts.clear(&workspace_id, &connection_id)).await
}
